pkg/querier/frontend/record_rule_substitution.go

```diff
@@ -0,0 +1,331 @@
+				lastHash = h
+				if err := qrrMap.LoadFromFile(event.Name); err != nil {
+					lastHash = ""
+					level.Error(log).Log("msg", "reloading config file for RecordRuleSubstitutionMiddleware", "err", err.Error())
+				}
+			}
+
+		case err := <-watcher.Errors:
+			if err != nil {
+				level.Error(log).Log("msg", "watching config file for RecordRuleSubstitutionMiddleware", "err", err)
+			}
+		}
+	}
+}
+
+type recordRuleSubstitution struct {
+	next   queryRangeHandler
+	qrrMap *RecordRuleSubstitutionConfig
+
+	configFilename string
+	watcher        *fsnotify.Watcher
+}
+
+func (rr recordRuleSubstitution) Do(ctx context.Context, r *QueryRangeRequest) (*APIResponse, error) {
+	reqs, err := rr.replaceQueryWithRecordingRule(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+
+	reqResps, err := doRequests(ctx, rr.next, reqs)
+	if err != nil {
+		return nil, err
+	}
+
+	resps := make([]*APIResponse, 0, len(reqResps))
+	for _, reqResp := range reqResps {
+		resps = append(resps, reqResp.resp)
+	}
+
+	return mergeAPIResponses(resps)
+}
+
+func (rr recordRuleSubstitution) replaceQueryWithRecordingRule(ctx context.Context, r *QueryRangeRequest) ([]*QueryRangeRequest, error) {
+	if rr.qrrMap == nil {
+		return []*QueryRangeRequest{r}, nil
+	}
+
+	org, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	rReplaced, maxModifiedAt, err := rr.qrrMap.ReplaceQueryWithRecordingRule(org, r)
+	if err != nil {
+		return nil, err
+	}
+
+	if rReplaced == nil {
+		return []*QueryRangeRequest{r}, nil
+	}
+
+	modifiedAtMillis := maxModifiedAt.UnixNano() / int64(time.Millisecond/time.Nanosecond)
+	modifiedAtMillis = (modifiedAtMillis / r.Step) * r.Step
+	if modifiedAtMillis > r.End {
+		// No recording rule exist for the query range.
+		return []*QueryRangeRequest{r}, nil
+	}
+	if modifiedAtMillis <= r.Start {
+		// Recording rule exists for entire query range.
+		return []*QueryRangeRequest{rReplaced}, nil
+	}
+
+	// Recording rule exists for partial query range.
+
+	alignedSafetyOffset := (safetyOffset / r.Step) * r.Step
+	r.End = modifiedAtMillis + alignedSafetyOffset
+
+	rReplaced.Start = r.End + r.Step
+	if rReplaced.Start >= rReplaced.End {
+		// Very less evaulations using the recording rule.
+		// This is also possible because of adding 'alignedSafetyOffset'.
+		return []*QueryRangeRequest{r}, nil
+	}
+
+	return []*QueryRangeRequest{r, rReplaced}, nil
+}
+
+// generaliseQuery returns the query by formatting it with PromQL printer.
+// This will generalise the query structure, and also check for errors in query.
+func generaliseQuery(q string) (string, error) {
+	expr, err := promql.ParseExpr(q)
+	if err != nil {
+		return "", err
+	}
+	return expr.String(), nil
+}
+
+// RecordRuleSubstitutionConfig gives thread safe access to
+// query to recording rule map of all organisations.
+type RecordRuleSubstitutionConfig struct {
+	qrrMap map[string][]QueryToRecordingRuleMap // (org id) -> its query-recording rule maps.
+	mtx    sync.RWMutex
```

Comment:
Do you need this extra struct?  Can these fields be embedded in recordRuleSubstitution

Reply:
We would want to have a single config entity so that any modifications to the config is reflected everywhere (all the requests). I see that a new `recordRuleSubstitution` would be created for every query, so I kept config as separate as reload becomes simple.